proposals/text/0000-evaluation-haystack-2.md

```diff
@@ -0,0 +1,223 @@
+```
+
+Or return some progress percentage and only in the end the final result.
+
+```
+def eval(runnable, inputs, expected_outputs):
+    outputs = []
+    total = len(inputs)
+    for i, input_ in enumerate(inputs):
+      output = runnable.run(input_)
+      outputs.append(output)
+      yield 100 * (i / total), None
+
+    yield 100, EvaluationResult(runnable, inputs, outputs, expected_outputs)
+```
+
+### `EvaluationResult`
+
+`EvaluationResult` won't contain partial metrics anymore as it does in v1.x but it will keep track of all the information used by `eval()`.
+This way we'll be able to save all the necessary information to a single file. That information will probably be:
+
+- Serialized Pipeline or Component
+- Inputs
+- Expected outputs
+- Actual outputs
+
+This data should be serializable to string so that it can be saved to file and loaded back whenever necessary.
+We shouldn't expect all input and output data to implement serialization methods like `to_dict` and `from_dict` like `Pipeline` and `component`s do, so we probably should find an alternative to handle serialization of types that don't. An unsafe option would be the use of `pickle`, but that's dangerous as it could lead to malicious code being executed.
+
+Given the above information we should be able to implement a single method to calculate predeterminated metrics or even custom ones.
+Known metrics could be defined as an enum to ease discoverability and documentation.
+
+```
+class Metric(Enum):
+    RECALL = "Recall"
+    MRR = "Mean Reciprocal Rank"
+    MAP = "Mean Average Precision"
+    EM = "Exact Match"
+    F1 = "F1"
+    SAS = "SemanticAnswerSimilarity"
+```
+
+The method to calculate metrics could look similar to this:
+
+```
+MetricsResult = Dict[str, Dict[str, float]]
+MetricCalculator = Callable[..., MetricResult]
+
+def calculate_metrics(self: EvaluationResult, metric: Union[Metric, MetricCalculator], **kwargs) -> MetricsResult:
+    # Verify if we're calculating a known metric
+    if metric == Metric.RECALL:
+      return self._calculate_recall(**kwargs)
+    elif metric == Metric.MRR:
+      return self._calculate_mrr(**kwargs)
+    # Other metrics...
+
+    # If it's not a known metric it must be a custom one
+    return metric(self, **kwargs)
+```
+
+This gives the users the flexibility to easily calculate metrics that we support but also use custom logic to calculate any kind of metric given the available data.
+Since users will need to save their calculated metrics to file we could create a simple `MetricResult` class that simply wraps the generated metrics dictionary, something similar:
+
+```
+class MetricResult(dict):
+    def save(self, file: Union[str, Path]):
+        # Dump info to file here
+```
+
+### Ease input specification
+
+As it can be annoying having to specify the inputs and outputs for evaluation since we need to specify both the `component` name and its input name we could make some assumption when running `Pipeline`s.
```

Comment:
I'm sorry but I had to read this sentence three times to understand it :sweat_smile: can we make it a bit simpler?

Reply:
I'll rephrase it. 👍